J2735-2024 Part 3 of 9: MAP, PSM, RTCM

iyourshaw opened the pull request:
Includes messages:

* MapData
* PersonalSafetyMessage
* RTCMcorrections

and classes they depend on in:

* AddGrpB
* AddGrpC
* REGION

Does not include MessageFrames.

The number of files added in this PR is about 200 to accommodate reviewing in a browser.

Includes unit tests.  Overall test coverage is 89%.  Coverage report:
![image](https://github.com/user-attachments/assets/9b49564e-7800-43ee-9e6c-87309290928a)

These POJOs are mostly quite similar to the ones in the previous PR.  

Includes some fixes to the following classes in the runtime library, and one new runtime class, that will be needed for messages in PR parts 4 and 6:

- NestedSequenceOfSerializer
- OpenTypeSerializer
- Asn1Bitstring
- Asn1OctetString
- RootSequenceOfSerializer (new)

iyourshaw committed:
MapData classes

iyourshaw committed:
PSM classes

iyourshaw committed:
RTCM

iyourshaw committed:
BSM classes

iyourshaw committed:
Revert "BSM classes"

This reverts commit 0b7f62a9d69550073b71f12d984b96738f2529e1.

iyourshaw committed:
Include tests

iyourshaw committed:
Bitstring test base class.  Remove rtcm message frame.

iyourshaw committed:
update script

iyourshaw committed:
runtime updates from main branch

iyourshaw committed:
Add documentation to update.sh

iyourshaw committed:
Resolve conflict

iyourshaw committed:
Remove unused imports

mcook42 commented on jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/serialization/RootSequenceOfSerializer.java:
```diff
@@ -0,0 +1,50 @@
+  public void serialize(Asn1SequenceOf<Asn1Type> sequenceOf, JsonGenerator jsonGenerator, SerializerProvider serializerProvider)
+      throws IOException {
+    if (serializerProvider instanceof XmlSerializerProvider) {
+      // XER
```
issue(blocking): this class isn't used in any of the tests or classes included in the repo so far. Let's we remove this class until we need it. I'm not comfortable introducing a hack into the code without a reason I can test/confirm myself.

iyourshaw replied:
OK.  It will be needed later to properly serialize one of the message types (ProbeDataConfig in Part6), which is a SEQUENCE-OF.  All the other message types are SEQUENCEs.  It is gone now but will be back then.

mcook42 commented on jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/serialization/OpenTypeSerializer.java:
```diff
@@ -65,9 +66,15 @@ public void serialize(T t, JsonGenerator jsonGenerator, SerializerProvider seria
+                var mapper = (XmlMapper)xmlGen.getCodec();
+                String itemXml = mapper.writeValueAsString(t);
+                xmlGen.writeRaw(itemXml);
+            }
```
question(blocking): in what scenario would we see the wrapper be null?

iyourshaw replied:
When the open type is a sequence-of annotated with RootSequenceOfSerializer, which will happen for ProbeDataConfigMessageFrame.  It is reverted until then.

mcook42 commented on jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1OctetString.java:
```diff
@@ -68,7 +68,10 @@ protected void validate(String aValue) throws Exception {
         // Size of hex format string can be 2 * byte size
-        boolean validLength = aValue.length() >= 2 * minLength && aValue.length() <= 2 * maxLength;
+        int trimmedLength = aValue.trim().length();
+        long minLength2 = 2L * minLength;
+        long maxLength2 = 2L * maxLength; // prevent 2*Integer.MAX_VALUE is negative
+        boolean validLength = (trimmedLength >= minLength2) && (trimmedLength <= maxLength2);
```
question(blocking): when would `2*Integer.MAX_VALUE` be negative?

iyourshaw replied:
Always when it assigned to an integer. Integers in Java are twos complement so in the following:

```java
int maxLength = Integer.MAX_VALUE;
int maxLength2 = 2 * maxLength;
```

maxLength2 wraps around and the test (size <= maxLength2) fails.  Assigning to a long prevents that bug.

This is needed because some OCTET STRINGS in the standard are unbounded, with the upper bound represented as Integer.MAX_LENGTH.

iyourshaw replied:
I added a regression test that fails if this is reverted.
[Asn1OctetStringTest.java](https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-3/jpo-asn-runtime/src/test/java/us/dot/its/jpo/asn/runtime/types/Asn1OctetStringTest.java)

iyourshaw committed:
Revert open type/RootSequenceOfSerializer

payneBrandon commented on jpo-asn-j2735-2024/src/main/java/us/dot/its/jpo/asn/j2735/r2024/REGION/Reg_MapData.java:
```diff
@@ -0,0 +1,56 @@
+    super(id, name);
+  }
+
+  public static class Reg_MapDataDeserializer extends ParameterizedTypeDeserializer<Reg_MapData> {
```
question: is this Reg_MapData usage intentionally excluding the generic type parameter?

iyourshaw replied:
Yes that is intentional.  The type parameter in the enclosing class can't be accessed from the static inner class, nor is it possible to pass a generic type to the inner type's constructor.

The part 2 PR contained 5 other examples exactly like this.

iyourshaw replied:
Added some notes on the overall design choices to the README, incl a specific note to hopefully clarify why all the raw types warnings are there:
https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-3/jpo-asn-j2735-2024/README.md#parameterized-types

payneBrandon commented on jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/serialization/NestedSequenceOfSerializer.java:
```diff
@@ -10,13 +10,16 @@
  * @author Ivan Yourshaw
  */
+@Slf4j
 public class NestedSequenceOfSerializer<T extends Asn1SequenceOf<?>> extends StdSerializer<T> {
```
question: is this planned for future use? I see no reference or unit tests thus far

iyourshaw replied:
Yes it will be needed for TIM messages (TravelerDataFrame) and others.

payneBrandon replied:
Are unit tests coming at that point as well then?

iyourshaw replied:
> Are unit tests coming at that point as well then?

Yes, there is a complete set of unit tests for all message types.  The TIM tests exercise this code.  They are here:

https://github.com/CDOT-CV/jpo-asn-pojos/tree/j2735-2024-part-4/jpo-asn-j2735-2024/src/test/java/us/dot/its/jpo/asn/j2735/r2024/TravelerInformation

payneBrandon commented on jpo-asn-j2735-2024/src/main/java/us/dot/its/jpo/asn/j2735/r2024/AddGrpB/LaneDataAttribute_addGrpBReg_LaneDataAttribute.java:
```diff
@@ -0,0 +1,67 @@
+import us.dot.its.jpo.asn.runtime.serialization.OpenTypeSerializer;
+
+@JsonRootName("Reg_LaneDataAttribute")
+@JsonDeserialize(using = None.class)
```
question: i assume this "None" here is to allow the custom ser/des in the `Reg_LaneDataAttribute` specified by `@Asn1ParameterizedTypes`? This was a bit convoluted to follow, is there some better documentation we can have to make this more clear?

iyourshaw replied:
Exactly, that "None" prevents an infinite recursion in the custom serializer on Reg_LaneDataAttribute.

It would be tricky (maybe possible but I feel too time-consuming) to add comments to the generated classes, but I can add a note in the README.

iyourshaw replied:
Added a note on this to the README here:

https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-3/jpo-asn-j2735-2024/README.md#custom-serializersdeserializers

payneBrandon replied:
Definitely helpful addition to that README. It's becoming much more robust!

iyourshaw committed:
Add regression test for octet string validation bug

payneBrandon commented on jpo-asn-j2735-2024/src/main/java/us/dot/its/jpo/asn/j2735/r2024/PersonalSafetyMessage/PersonalClusterRadius.java:
```diff
@@ -0,0 +1,54 @@
+import us.dot.its.jpo.asn.runtime.types.Asn1Integer;
+
+@JsonDeserialize(using = PersonalClusterRadius.PersonalClusterRadiusDeserializer.class)
+public class PersonalClusterRadius extends Asn1Integer {
```
question (non-blocking): is there any way to include the J2735 documentation for these types in the future? For instance, we set this radius to 0-100, but from a coding standpoint that isn't defined as to what "radius" is in terms of. Non-blocking as this may be much more work, but a future effort to add in documentation would be phenomenal for developer understanding

iyourshaw replied:
That is a great idea, and certainly would be convenient. It may be possible to add that to the generated classes automatically, but best left as a future task IMO, since it may be a tricky and/or time-consuming to implement.  Another concern might be that embedding a lot of verbatim text from the specification in this code may run afoul of the copyright on the spec.  "I am not a lawyer" so am not really sure.

payneBrandon replied:
Possible future enhancement then. I figured this wouldn't fit in this initial minimal viable product version, but wanted to bring it up to keep in mind for future improvements.

payneBrandon commented on jpo-asn-j2735-2024/src/main/java/us/dot/its/jpo/asn/j2735/r2024/MapData/MapData.java:
```diff
@@ -0,0 +1,106 @@
+  private SequenceOfRegional regional;
+
+  @JsonInclude(Include.NON_NULL)
+  @java.lang.SuppressWarnings("rawtypes")
```
question: I've seen this suppress warning in a few locations - is there plans in the future to handle the underlying cause?

iyourshaw replied:
Please see the discussion here where that SuppressWarnings was added:
https://github.com/CDOT-CV/jpo-asn-pojos/pull/13#pullrequestreview-2717824855

The underlying cause is that Java's generic's implementation is fundamentally awkward to work with.  

One workaround would be to not use generics at all here, but just object inheritance, but that would lose us expressiveness.  For example, for MessageFrames, being able to use:
```java
MessageFrame<MapData>
MessageFrame<BasicSafetyMessage>
```
etc. is IMO a good and expressive representation of what the ASN.1 parameterized types are intended to be, which is a set of various types that plug into a "hole" in another type in a modular and extensible fashion.  Eliminating the generic type could work, but would lose that expressiveness of the intent of the standard, so I would rather not.

Another workaround may be to re-work the base classes using something like this hack that the Jackson library's StdSerializer constructor uses:
[Jackson StdSerializer](https://www.javadoc.io/doc/com.fasterxml.jackson.core/jackson-databind/latest/com/fasterxml/jackson/databind/ser/std/StdSerializer.html#StdSerializer-java.lang.Class-boolean-)

```
StdSerializer

    protected StdSerializer(Class<?> t,  boolean dummy)

    Alternate constructor that is (alas!) needed to work around kinks of generic type handling
```
Or it could be possible to have a base type constructor accepting `java.lang.Type` instead of `Class<T>` in conjunction with using `java.reflect.ParameterizedType` to get the `.class` for a generic type with captured type parameter `<?>` to pass to it. That sort of thing would add complexity and require some testing and experimentation to get right.

Any of those approaches would be workarounds that don't address the underlying cause though, so there are no plans unless I'm missing something and someone has a better idea that is simple and fast to implement.

iyourshaw replied:
Added some notes on the overall design choices to the README, incl a specific note to hopefully clarify why all the raw types warnings are there:
https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-3/jpo-asn-j2735-2024/README.md#parameterized-types

payneBrandon replied:
Thanks for the description, and definitely thanks for updating the README! Having these design choices and explanations there will be extremely helpful for developers in the future who use this code.

payneBrandon reviewed (changes requested):
Looking good so far, a few questions I've come across

mcook42 committed:
refactor: delegate from default constructor in Asn1OctetString to prevent drift

mcook42 committed:
refactor: Asn1OctetString for better validation and clarity

Enhanced validation logic for hex strings and byte arrays with size constraints. Improved method documentation and added null checks to ensure robustness. Introduced utility methods for consistent length verification.

mcook42 committed:
chore: move Asn1OctetString constructors to top of file

mcook42 committed:
refactor: Asn1Bitstring constructors and improve exceptions

Added detailed Javadoc comments for constructors to enhance clarity. Introduced an overloaded constructor for consistent size handling. Improved exception messages to provide more context when input strings are too short.

mcook42 committed:
chore: update exception message to represent discrepancy between sizes

mcook42 reviewed (approved):
suggestion(non-blocking): i've made some changes over in https://github.com/CDOT-CV/jpo-asn-pojos/pull/22 that I found helpful while reviewing this change set.

overall, looks good enough to me

iyourshaw committed:
Add some notes to README

iyourshaw committed:
edit readme

iyourshaw committed:
edit readme

iyourshaw committed:
edit readme

iyourshaw committed:
Merge branch 'j2735-2024-part-3' into mcook42/suggest/j2735-2024-part-3

iyourshaw committed:
Add tests for deserializing octet string xml with various types of whitespace

mcook42 committed:
Revert "Add tests for deserializing octet string xml with various types of whitespace"

This reverts commit 28a2dbcdfebf73d441223e5109415415b924ea59.

mcook42 committed:
chore: trim hexString before validating size is multiple of 2

iyourshaw committed:
Merge pull request #22 from CDOT-CV/mcook42/suggest/j2735-2024-part-3

Suggested updates found while reviewing Part 3

payneBrandon reviewed (approved):
lgtm!
